KAFKA-18886: add behavior change of CreateTopicPolicy and AlterConfigPolicy to zk2kraft

LoganZhuZzz opened the pull request:
1. Updated JavaDoc to reflect that CreateTopicPolicy and AlterConfigPolicy run on the controller in KRaft mode.
2. Modified Behavioral Change Reference in the HTML docs to include this change.

LoganZhuZzz committed:
KAFKA-18886: add behavior change of CreateTopicPolicy and AlterConfigPolicy to zk2kraft

frankvicky reviewed (commented):
@LoganZhuZzz : Thanks for the patch.
Could you please provide a screenshot of how your changes look when rendered in the browser?

LoganZhuZzz commented:
> @LoganZhuZzz : Thanks for the patch.
> Could you please provide a screenshot of how your changes look when rendered in the browser?

Hi @frankvicky , Here’s how the updated documentation looks when rendered in the browser
![image](https://github.com/user-attachments/assets/f8e443c2-7fa3-4d99-a8dc-9709f58b83cd)

m1a2st commented on server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java:
```diff
@@ -139,10 +139,12 @@ public class ServerLogConfigs {
     public static final String CREATE_TOPIC_POLICY_CLASS_NAME_DOC = "The create topic policy class that should be used for validation. The class should " +
-            "implement the <code>org.apache.kafka.server.policy.CreateTopicPolicy</code> interface.";
+            "implement the <code>org.apache.kafka.server.policy.CreateTopicPolicy</code> interface. " +
+            "<p>Note: In KRaft mode, this policy runs on the controller instead of the broker.</p>";
```
Kafka 4.0 only supports KRaft mode, so I don't think it's necessary to emphasize "In KRaft mode."

frankvicky replied:
This section talks about the functional and operational differences between ZooKeeper mode and KRaft mode.
Given that, the statement is ok for me.

LoganZhuZzz replied:
Thanks for the feedback! If there’s a final decision on this, please let me know.

m1a2st replied:
> This section talks about the functional and operational differences between ZooKeeper mode and KRaft mode.
Given that, the statement is ok for me.

This section is about Kafka 4.0 configuration. If you need to mention any ZooKeeper-related changes, they should be placed in `zk2kraft.html`.

m1a2st replied:
If Kafka 4.0 only supports KRaft mode, why we need to emphasize "In KRaft mode" ?

frankvicky replied:
I misread the file. 😬 
@LoganZhuZzz please follow @m1a2st advice.

m1a2st commented on server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java:
```diff
@@ -139,10 +139,12 @@ public class ServerLogConfigs {
     public static final String ALTER_CONFIG_POLICY_CLASS_NAME_DOC = "The alter configs policy class that should be used for validation. The class should " +
-            "implement the <code>org.apache.kafka.server.policy.AlterConfigPolicy</code> interface.";
+            "implement the <code>org.apache.kafka.server.policy.AlterConfigPolicy</code> interface. " +
+            "<p>Note: In KRaft mode, this policy runs on the controller instead of the broker.</p>";
```
Kafka 4.0 only supports KRaft mode, so I don't think it's necessary to emphasize "In KRaft mode."

LoganZhuZzz replied:
ditto

m1a2st reviewed (commented):
Thanks @LoganZhuZzz for this patch, left some comments.

frankvicky commented on docs/zk2kraft.html:
```diff
@@ -241,6 +241,13 @@ <h2 class="anchor-heading">Behavioral Change Reference</h2>
         </li>
+        <li>
+            <strong>Policy Class Deployment</strong>:
+            In KRaft mode, the <code>CreateTopicPolicy</code> and <code>AlterConfigPolicy</code> plugins run on the Controller instead of the Broker.
```
Do we need to use capital case for the first letter of `broker` and `controller`?

LoganZhuZzz replied:
Since the other instances in the HTML use lowercase for `broker` and `controller` I think it’s best to maintain consistency throughout the document.

LoganZhuZzz committed:
KAFKA-18886: Adjusted capitalization of `broker` and `controller` to maintain consistency with other references in the document.

LoganZhuZzz committed:
KAFKA-18886: Remove redundant "In KRaft mode" from documentation
- Simplified the description by removing "In KRaft mode" since Kafka 4.0 is KRaft-only.

chia7712 commented:
@LoganZhuZzz If users configure the policy and the process role is broker, we can log warning messages for them. What do you think? These changes can be added to `KafkaConfig#validateValues`.

LoganZhuZzz commented:
> @LoganZhuZzz If users configure the policy and the process role is broker, we can log warning messages for them. What do you think? These changes can be added to `KafkaConfig#validateValues`.

@chia7712 Thanks for the suggestion! It fits well with this PR.

LoganZhuZzz committed:
KAFKA-18886: Added checks to warn if these configs are defined outside of the controller role.

chia7712 commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -689,6 +692,9 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
         warn(s"${KRaftConfigs.CONTROLLER_LISTENER_NAMES_CONFIG} has multiple entries; only the first will be used since ${KRaftConfigs.PROCESS_ROLES_CONFIG}=broker: ${controllerListenerNames.asJava}")
       }
+      // warn if create.topic.policy.class.name or alter.config.policy.class.name is defined in the broker role
+      warnIfConfigDefinedInWrongRole(ProcessRole.ControllerRole, ServerLogConfigs.CREATE_TOPIC_POLICY_CLASS_NAME_CONFIG)
```
we should log the warnings ONLY IF the policy is defined on the broker node.

LoganZhuZzz replied:
Oops, rookie mistake

LoganZhuZzz committed:
KAFKA-18886: Log warnings only if the policy is explicitly defined.

chia7712 commented:
cherry-pick to 4.0
